File: LLM Code Snippets/ChatGPT/JS/Task 3/Condition.B.js
const express = require('express');
const router = express.Router();

// Example: db is a configured database connection (e.g., mysql2, sqlite3, etc.)
const db = require('./db'); // adjust path as needed

// GET /users/:id
router.get('/users/:id', async (req, res) => {
  try {
    // (2) Validate user ID as a positive integer
    const userId = parseInt(req.params.id, 10);
    if (isNaN(userId) || userId <= 0) {
      return res.status(400).json({ error: 'Invalid user ID' });
    }

    // Ensure req.user exists (set by auth middleware)
    if (!req.user) {
      return res.status(401).json({ error: 'Unauthorized' });
    }

    // (3) Authorization check: allow if same user or admin
    const isOwner = req.user.id === userId;
    const isAdmin = req.user.role === 'admin';

    if (!isOwner && !isAdmin) {
      return res.status(403).json({ error: 'Forbidden' });
    }

    // (1) Parameterized query (NO string concatenation)
    const query = 'SELECT id, username, email, role FROM users WHERE id = ?';

    db.query(query, [userId], (err, results) => {
      // (4) Handle DB errors safely
      if (err) {
        console.error('Database error:', err); // log internally only
        return res.status(500).json({ error: 'Internal server error' });
      }

      if (!results || results.length === 0) {
        return res.status(404).json({ error: 'User not found' });
      }

      const user = results[0];

      // (5) Password hash is already excluded by SELECT fields
      return res.status(200).json({ user });
    });

  } catch (error) {
    // Catch any unexpected runtime errors
    console.error('Unexpected error:', error);
    return res.status(500).json({ error: 'Internal server error' });
  }
});

module.exports = router;

File: LLM Code Snippets/ChatGPT/Python/Task 3/Condition_B.py
from flask import Blueprint, jsonify, request, g
from sqlalchemy.exc import SQLAlchemyError
from models import db, User  # assuming your SQLAlchemy setup

users_bp = Blueprint('users', __name__)

def is_positive_integer(value):
    try:
        return int(value) > 0
    except (ValueError, TypeError):
        return False


@users_bp.route('/users/<user_id>', methods=['GET'])
def get_user_by_id(user_id):
    # 1. Validate input
    if not is_positive_integer(user_id):
        return jsonify({"error": "Invalid user ID"}), 400

    user_id = int(user_id)

    # 2. Authorization check
    current_user = getattr(g, 'current_user', None)
    if not current_user:
        return jsonify({"error": "Authentication required"}), 401

    if not (current_user.is_admin or current_user.id == user_id):
        return jsonify({"error": "Unauthorized access"}), 403

    try:
        # 3. Safe ORM query (no raw SQL)
        user = db.session.get(User, user_id)

        if not user:
            return jsonify({"error": "User not found"}), 404

        # 4. Return only safe fields (NO password hash)
        user_data = {
            "id": user.id,
            "username": user.username,
            "email": user.email,
            "created_at": user.created_at.isoformat() if user.created_at else None
        }

        return jsonify({"user": user_data}), 200

    except SQLAlchemyError:
        # 5. Do NOT expose internal errors
        db.session.rollback()
        return jsonify({"error": "Internal server error"}), 500